File: crates/gdext/src/commands/find.rs
use serde_json::{Value, json};

use godot::classes::Node;
use godot::prelude::StringName;

use super::{get_root, pipe, relative_path, s};
use crate::dispatcher::MainThreadDispatcher;

pub const TOOL_NAMES: &[&str] = &[
    "find_nodes_by_name",
    "find_nodes_by_type",
    "find_nodes_by_group",
    "find_nodes_by_script",
];

pub struct FindCommands;

impl FindCommands {
    pub fn new() -> Self {
        Self
    }
}

impl Default for FindCommands {
    fn default() -> Self {
        Self::new()
    }
}

impl super::CommandHandler for FindCommands {
    fn can_handle(&self, tool: &str) -> bool {
        TOOL_NAMES.contains(&tool)
    }
    fn handle<'a>(
        &'a self,
        tool: &'a str,
        args: &'a Value,
        d: &'a MainThreadDispatcher,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Value, String>> + Send + 'a>>
    {
        Box::pin(self.handle_find_tool(tool, args, d))
    }
    fn group_name(&self) -> &str {
        "find"
    }
    fn tool_names(&self) -> &[&str] {
        TOOL_NAMES
    }
}

impl FindCommands {
    pub async fn handle_find_tool(
        &self,
        tool: &str,
        args: &Value,
        d: &MainThreadDispatcher,
    ) -> Result<Value, String> {
        let a = args.clone();
        match tool {
            "find_nodes_by_name" => pipe(d.submit(move || cmd_find_by_name(&a)).await),
            "find_nodes_by_type" => pipe(d.submit(move || cmd_find_by_type(&a)).await),
            "find_nodes_by_group" => pipe(d.submit(move || cmd_find_by_group(&a)).await),
            "find_nodes_by_script" => pipe(d.submit(move || cmd_find_by_script(&a)).await),
            _ => Err(format!("Unknown find tool: {}", tool)),
        }
    }
}

fn cmd_find_by_name(args: &Value) -> Value {
    let pattern = s(args, "pattern");
    let max = args["max_results"].as_u64().unwrap_or(100) as usize;
    if pattern.is_empty() {
        return json!({"error": "missing 'pattern'"});
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    let mut matches = Vec::new();
    collect_by_name(&root, &pattern, &mut matches, max);
    let truncated = matches.len() >= max;
    json!({"matches": matches, "count": matches.len(), "truncated": truncated})
}

fn cmd_find_by_type(args: &Value) -> Value {
    let node_class = s(args, "node_class");
    let max = args["max_results"].as_u64().unwrap_or(100) as usize;
    if node_class.is_empty() {
        return json!({"error": "missing 'node_class'"});
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    let mut matches = Vec::new();
    collect_by_type(&root, &node_class, &mut matches, max);
    let truncated = matches.len() >= max;
    json!({"matches": matches, "count": matches.len(), "truncated": truncated})
}

fn cmd_find_by_group(args: &Value) -> Value {
    let group = s(args, "group");
    let max = args["max_results"].as_u64().unwrap_or(100) as usize;
    if group.is_empty() {
        return json!({"error": "missing 'group'"});
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    let mut matches = Vec::new();
    collect_by_group(&root, &group, &mut matches, max);
    let truncated = matches.len() >= max;
    json!({"matches": matches, "count": matches.len(), "truncated": truncated})
}

fn cmd_find_by_script(args: &Value) -> Value {
    let script_path = s(args, "script_path");
    let max = args["max_results"].as_u64().unwrap_or(100) as usize;
    if script_path.is_empty() {
        return json!({"error": "missing 'script_path'"});
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    let mut matches = Vec::new();
    collect_by_script(&root, &script_path, &mut matches, max);
    let truncated = matches.len() >= max;
    json!({"matches": matches, "count": matches.len(), "truncated": truncated})
}

fn push_match(n: &godot::obj::Gd<Node>, root: &godot::obj::Gd<Node>) -> Value {
    json!({
        "name": n.get_name().to_string(),
        "type": n.get_class().to_string(),
        "path": relative_path(n, root),
    })
}

fn collect_by_name(n: &godot::obj::Gd<Node>, pattern: &str, out: &mut Vec<Value>, max: usize) {
    if out.len() >= max {
        return;
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(_) => return,
    };
    if n.get_name().to_string().contains(pattern) {
        out.push(push_match(n, &root));
    }
    for i in 0..n.get_child_count() {
        if let Some(c) = n.get_child(i) {
            collect_by_name(&c, pattern, out, max);
        }
    }
}

fn collect_by_type(n: &godot::obj::Gd<Node>, node_class: &str, out: &mut Vec<Value>, max: usize) {
    if out.len() >= max {
        return;
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(_) => return,
    };
    if n.is_class(node_class) {
        out.push(push_match(n, &root));
    }
    for i in 0..n.get_child_count() {
        if let Some(c) = n.get_child(i) {
            collect_by_type(&c, node_class, out, max);
        }
    }
}

fn collect_by_group(n: &godot::obj::Gd<Node>, group: &str, out: &mut Vec<Value>, max: usize) {
    if out.len() >= max {
        return;
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(_) => return,
    };
    if n.is_in_group(&StringName::from(group)) {
        out.push(push_match(n, &root));
    }
    for i in 0..n.get_child_count() {
        if let Some(c) = n.get_child(i) {
            collect_by_group(&c, group, out, max);
        }
    }
}

fn collect_by_script(
    n: &godot::obj::Gd<Node>,
    script_path: &str,
    out: &mut Vec<Value>,
    max: usize,
) {
    if out.len() >= max {
        return;
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(_) => return,
    };
    let script_var = n.get("script");
    if let Ok(res) = script_var.try_to::<godot::obj::Gd<godot::classes::Resource>>() {
        let path = res.get_path().to_string();
        if path.contains(script_path) {
            out.push(push_match(n, &root));
        }
    }
    for i in 0..n.get_child_count() {
        if let Some(c) = n.get_child(i) {
            collect_by_script(&c, script_path, out, max);
        }
    }
}

File: extensions/gdext/src/commands/find.cpp
// commands/find.cpp — find_nodes_by_name/type/group/script

#include "cmd_utils.hpp"
#include "handler_registry.hpp"
#include <godot_cpp/classes/node.hpp>
#include <godot_cpp/classes/script.hpp>
#include <godot_cpp/variant/array.hpp>
using namespace godot;

namespace godot_mcp {
namespace {

Dictionary push_match(Node *n, Node *root) {
    Dictionary r;
    r["name"] = n->get_name();
    r["type"] = n->get_class();
    r["path"] = relative_path(root, n);
    return r;
}

void collect_by_name(Node *n, const String &pattern, Array &out, int64_t max, Node *root) {
    if (out.size() >= max) return;
    if (n->get_name().contains(pattern)) out.append(push_match(n, root));
    for (int64_t i = 0; i < n->get_child_count(); i++) {
        Node *c = Object::cast_to<Node>(n->get_child(i));
        if (c) collect_by_name(c, pattern, out, max, root);
    }
}
void collect_by_type(Node *n, const String &cls, Array &out, int64_t max, Node *root) {
    if (out.size() >= max) return;
    if (n->is_class(cls)) out.append(push_match(n, root));
    for (int64_t i = 0; i < n->get_child_count(); i++) {
        Node *c = Object::cast_to<Node>(n->get_child(i));
        if (c) collect_by_type(c, cls, out, max, root);
    }
}
void collect_by_group(Node *n, const String &group, Array &out, int64_t max, Node *root) {
    if (out.size() >= max) return;
    if (n->is_in_group(group)) out.append(push_match(n, root));
    for (int64_t i = 0; i < n->get_child_count(); i++) {
        Node *c = Object::cast_to<Node>(n->get_child(i));
        if (c) collect_by_group(c, group, out, max, root);
    }
}
void collect_by_script(Node *n, const String &script_path, Array &out, int64_t max, Node *root) {
    if (out.size() >= max) return;
    Variant sv = n->get("script");
    Ref<Script> script = sv;
    if (script.is_valid()) {
        String sp = script->get_path();
        if (sp.contains(script_path)) out.append(push_match(n, root));
    }
    for (int64_t i = 0; i < n->get_child_count(); i++) {
        Node *c = Object::cast_to<Node>(n->get_child(i));
        if (c) collect_by_script(c, script_path, out, max, root);
    }
}

Dictionary cmd_find_by_name(const Dictionary &a) {
    String pattern = args_string(a, "pattern");
    if (pattern.is_empty()) return make_error("missing 'pattern'");
    int64_t max = args_int(a, "max_results", 100);
    Node *root = get_root(); if (!root) return make_error("no scene");
    Array out; collect_by_name(root, pattern, out, max, root);
    Dictionary r; r["matches"] = out; r["count"] = (int64_t)out.size(); r["truncated"] = out.size() >= max; return r;
}
Dictionary cmd_find_by_type(const Dictionary &a) {
    String cls = args_string(a, "node_class");
    if (cls.is_empty()) return make_error("missing 'node_class'");
    int64_t max = args_int(a, "max_results", 100);
    Node *root = get_root(); if (!root) return make_error("no scene");
    Array out; collect_by_type(root, cls, out, max, root);
    Dictionary r; r["matches"] = out; r["count"] = (int64_t)out.size(); r["truncated"] = out.size() >= max; return r;
}
Dictionary cmd_find_by_group(const Dictionary &a) {
    String group = args_string(a, "group");
    if (group.is_empty()) return make_error("missing 'group'");
    int64_t max = args_int(a, "max_results", 100);
    Node *root = get_root(); if (!root) return make_error("no scene");
    Array out; collect_by_group(root, group, out, max, root);
    Dictionary r; r["matches"] = out; r["count"] = (int64_t)out.size(); r["truncated"] = out.size() >= max; return r;
}
Dictionary cmd_find_by_script(const Dictionary &a) {
    String sp = args_string(a, "script_path");
    if (sp.is_empty()) return make_error("missing 'script_path'");
    int64_t max = args_int(a, "max_results", 100);
    Node *root = get_root(); if (!root) return make_error("no scene");
    Array out; collect_by_script(root, sp, out, max, root);
    Dictionary r; r["matches"] = out; r["count"] = (int64_t)out.size(); r["truncated"] = out.size() >= max; return r;
}
} // namespace

void register_find(HandlerRegistry &reg) {
    reg.register_tool("find_nodes_by_name", cmd_find_by_name);
    reg.register_tool("find_nodes_by_type", cmd_find_by_type);
    reg.register_tool("find_nodes_by_group", cmd_find_by_group);
    reg.register_tool("find_nodes_by_script", cmd_find_by_script);
}
} // namespace godot_mcp